test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,293 @@
+# The purpose of this test is to check that we have implementation parity between
+# a Python `torch.nn` module and its corresponding C++ `torch::nn` module. Concretely,
+# this test does the following:
+#
+# 1. Get a test params dict from common_nn.py, run forward and backward on the
+# Python module created using the test params.
+#
+# 2. Serialize the Python module's parameters / buffers and its forward input
+# arguments, deserialize them in C++ and load them into the C++ module.
+#
+# 3. Run the same forward and backward passes on the C++ module, and serialize
+# the C++ module's forward output and backward gradients.
+#
+# 4. Compare Python/C++ module's forward output and backward gradients. If they
+# are the same, then we have implementation parity between Python/C++ module.
+
+import tempfile
+import shutil
+from string import Template
+import types
+
+import torch
+from cpp_api_parity.utils import TorchNNModuleTestParams, TORCH_NN_COMMON_TEST_HARNESS, \
+    compile_cpp_code_inline, set_python_tensors_requires_grad, move_python_tensors_to_device, \
+    add_test, compute_cpp_args_construction_stmts_and_forward_arg_symbols, serialize_arg_dict_as_script_module, \
+    compute_arg_dict, decorate_test_fn, compute_temp_file_path, generate_error_msg
+from cpp_api_parity import torch_nn_modules
+
+# Expected substitutions:
+#
+# ${module_variant_name}
+# ${module_qualified_name}
```

Comment:
What's the difference between this and `module_variant_name`?

Reply:
`module_qualified_name` is something like `torch::nn::Linear` (with the `torch::nn::` qualifier), `module_variant_name` is something like `Linear_no_bias_cpu`. I just added some comments to explain this better :D

Reply:
thanks!